docs/cryptodoc/src/05_10_cmce.rst

```diff
@@ -0,0 +1,469 @@
+where :math:`\alpha_i` are the first :math:`n` field ordering elements and
+:math:`g` is the Goppa polynomial.
+
+Next, the error locator polynomial :math:`\sigma` is computed using the
+Berlekamp-Massey algorithm on the syndrome. The resulting polynomial has a
+particular property concerning the error vector :math:`e` we search for.
+Specifically, :math:`\sigma(\alpha_i) = 0` if and only if
+:math:`e_i = 1`. By evaluating :math:`\sigma` at :math:`\alpha_0,\dots,\alpha_{n-1}`,
+we can reconstruct the error vector :math:`e`.
+
+To ensure accurate decoding, Botan follows the recommendation in Section 6.3 of
+the implementation guide of [CMCE-R4]_. It computes the syndrome for the error
+vector :math:`e` and compares it with the syndrome for :math:`\nu`. If the
+comparison holds and the weight of :math:`e` is equal to :math:`t`, we consider
+the decoding successful. Otherwise, it is flagged as a failure.
+
+It is worth noting that the syndrome computation, Berlekamp-Massey algorithm, and
+locator polynomial evaluation are implemented in constant time. Additionally,
+the checks for the weight of :math:`e` and the syndrome comparison are designed
+to avoid early abortion if any check fails.
+
+
+.. _pubkey/cmce/params:
+
+Parameters
+^^^^^^^^^^
+
+The ``Classic_McEliece_Parameter_Set`` enum contains all instances of the
+Classic McEliece algorithm listed in Table :ref:`Supported Classic McEliece
+Parameter Sets <pubkey/cmce/parameter_table>`. These parameter sets serve as
+the basis for deriving all the necessary parameters used within the algorithm.
+They are collected in a ``Classic_McElice_Parameters`` object, which includes
+all the parameters defined in the specifications [CMCE-ISO]_ and [CMCE-R4]_.
+This object is passed to all algorithm components, ensuring consistent
+parameter usage.
+
+
+.. _pubkey/cmce/key_gen:
+
+
+Key Generation
+--------------
+
+Classic McEliece key generation follows Section 8.3 of [CMCE-ISO]_ and is
+implemented within the ``Classic_McEliece_KeyPair_Internal`` class. It works as
+follows:
+
+.. admonition:: Classic McEliece Key Generation
+
+   **Input:**
+
+   -  ``rng``: random number generator
+
+   **Output:**
+
+   -  ``SK``, ``PK``: private and public key
+
+   **Steps:**
+
+   1. Generate a random value ``seed`` using ``rng``
+   2. ``s, ordering_bits, irreducible_bits, next_seed = PRF(seed)``
+   3. | Create a field ordering ``field_ordering`` using ``ordering_bits``
+      | On failure, set ``seed = next_seed`` and go to step 2
+
+   4. | Create a monic irreducible polynomial ``g`` using ``irreducible_bits``
+      | On failure, set ``seed = next_seed`` and go to step 2
+
+   5. | Create a parity check matrix in systematic form ``H = (I_mt | T)`` using ``field_ordering`` and ``g``. Meanwhile, compute the column selection ``c``
+      | On failure, set ``seed = next_seed`` and go to step 2
```

Comment:
Should we mention somewhere how failures are represented in our implementation?

Reply:
Do you mean by using `std::optional`? IMO, that's too much technically detailed.